tephra-api/src/main/java/co/cask/tephra/Transaction.java

```diff
@@ -108,18 +157,60 @@ public long getFirstShortInProgress() {
     return firstShortInProgress;
   }
 
+  /**
+   * Returns true if the given version corresponds to a transaction that was in-progress at the time this transaction
+   * started.
+   */
   public boolean isInProgress(long version) {
     return Arrays.binarySearch(inProgress, version) >= 0;
   }
 
+  /**
+   * Returns true if the given version is present in one of the arrays of excluded versions (in-progress and
+   * invalid transactions).
+   */
   public boolean isExcluded(long version) {
     return Arrays.binarySearch(inProgress, version) >= 0
       || Arrays.binarySearch(invalids, version) >= 0;
   }
 
+  /**
+   * Returns true if the the given version corresponds to one of the checkpoint versions in the current
+   * transaction.
+   */
+  public boolean isCheckpoint(long version) {
+    return Arrays.binarySearch(checkpointWritePointers, version) >= 0;
+  }
+
+  /**
+   * Returns whether or not the given version should be visible to the current transaction.  A version will be visible
+   * if it was successfully committed prior to the current transaction starting, or was written by the current
+   * transaction (using either the current write pointer or the write pointer from a prior checkpoint).
+   *
+   * @param version the data version to check for visibility
+   * @return true if the version is visible, false if it should be hidden (filtered)
+   *
+   * @see #isVisible(long, boolean) to exclude the current write pointer from visible versions.  This method always
+   *      includes the current write pointer.
+   */
   public boolean isVisible(long version) {
+    return isVisible(version, true);
+  }
+
+  /**
+   * Returns whether or not the given version should be visible to the current transaction.  A version will be visible
+   * if it was successfully committed prior to the current transaction starting, or was written by the current
+   * transaction (using either the current write pointer or the write pointer from a prior checkpoint).
+   *
+   * @param version the data version to check for visibility
+   * @param excludeCurrentWritePointer whether writes from the current write pointer should be visible
+   * @return true if the version is visible, false if it should be hidden (filtered)
+   */
+  public boolean isVisible(long version, boolean excludeCurrentWritePointer) {
```

Comment:
Thanks for the explanation, Gary. Yes, this will allow standard SQL semantics (i.e. a statement does not see its own writes) for relational database layers such as Phoenix.